File: .cache/score.mjs
import fs from 'fs';

const markets = JSON.parse(fs.readFileSync('.cache/cg-markets.json', 'utf8'));
const trending = JSON.parse(fs.readFileSync('.cache/cg-trending.json', 'utf8'));
const dex = JSON.parse(fs.readFileSync('.cache/dex-trending.json', 'utf8'));

const BTC_7D = -4.596;
const ETH_7D = -7.413;

const DEDUP = new Set(['KAIA','TRAC','BSB','INJ','NEAR','HYPE','CHZ','LIT','VVV','DASH','ZEC','PENGU','GRASS','ONDO','WLD','BEAT']);

const SKIP_PREFIX = ['USD','EUR','GBP','JPY','BRZ'];
const SKIP_EXACT = new Set([
  'USDT','USDC','DAI','USDE','FDUSD','TUSD','PYUSD','BUSD','GUSD','LUSD','FRAX','SUSDS','USDS','USDL','USD1','RUSD','AUSD','USDT0','USDD','USTC','USDP','XAUT','PAXG','XAU','USD0','USDX','USDB','SUSDE','USYC','BUIDL','OUSG','USR','USDH','USDM','GHO','USDF','USDR','SUSD','EURC','EURS','EURT','USDA','SDAI',
  'WBTC','WETH','WBNB','WSOL','WMATIC','WAVAX','WTRX','WBETH','WEETH','WSTETH','STETH','METH','RETH','CBETH','LSETH','CBBTC','TBTC','LBTC','SOLVBTC','BSOL','JITOSOL','JUPSOL','MSOL','BNSOL','ETHX','OSETH','RSETH','STAFI','RSWETH','EZETH','PUFETH','RBNB','BBTC','WHBAR','SBTC','XSOLVBTC',
  'U'
]);

function isStableOrWrap(sym) {
  if (SKIP_EXACT.has(sym)) return true;
  for (const p of SKIP_PREFIX) if (sym.startsWith(p) && sym.length <= 5) return true;
  return false;
}

const trendingSet = new Set(trending.coins.map(c => c.item.symbol.toUpperCase()));

const dexSet = new Set();
if (dex.pairs) {
  for (const p of dex.pairs.slice(0, 40)) {
    if (p.baseToken && p.baseToken.symbol) dexSet.add(p.baseToken.symbol.toUpperCase());
  }
}

const scored = [];
for (const m of markets) {
  if (!m.symbol) continue;
  const sym = m.symbol.toUpperCase();
  if (isStableOrWrap(sym)) continue;
  const mcap = m.market_cap || 0;
  if (mcap < 20_000_000) continue;
  const vol = m.total_volume || 0;
  const vmc = mcap > 0 ? vol / mcap : 0;
  const p24 = m.price_change_percentage_24h_in_currency ?? m.price_change_percentage_24h ?? 0;
  const p7 = m.price_change_percentage_7d_in_currency ?? 0;

  let s = 0;
  const sigs = [];
  if (p24 > 0) { s += 1; sigs.push('24h+'); }
  if (p7 > 0) { s += 1; sigs.push('7d+'); }
  if (p24 > 5 && p7 > 5) { s += 2; sigs.push('both>+5'); }
  if (trendingSet.has(sym)) { s += 2; sigs.push('trending'); }
  if (vmc >= 0.20) { s += 3; sigs.push('vmc>=0.20(' + vmc.toFixed(2) + ')'); }
  else if (vmc >= 0.10) { s += 2; sigs.push('vmc>=0.10(' + vmc.toFixed(2) + ')'); }
  if (p7 > BTC_7D && p7 > ETH_7D) { s += 2; sigs.push('RS>BTC&ETH 7d'); }
  if (dexSet.has(sym)) { s += 1; sigs.push('dex-confirm'); }

  const dedup = DEDUP.has(sym);
  scored.push({ sym, name: m.name, price: m.current_price, mcap, vol, vmc, p24, p7, score: s, sigs, dedup });
}

scored.sort((a, b) => b.score - a.score || b.vmc - a.vmc);

console.log('=== TOP 30 (excluding dedup) ===');
let n = 0;
for (const r of scored) {
  if (r.dedup) continue;
  if (n++ >= 30) break;
  console.log([r.score, r.sym, r.name, '$' + r.price, '24h=' + r.p24.toFixed(2), '7d=' + r.p7.toFixed(2), 'mc=$' + (r.mcap/1e6).toFixed(0) + 'M', 'vol=$' + (r.vol/1e6).toFixed(0) + 'M', 'vmc=' + r.vmc.toFixed(3), r.sigs.join('|')].join('\t'));
}
console.log('\n=== TOP 15 OVERALL (incl dedup) ===');
for (const r of scored.slice(0, 15)) {
  console.log([r.score, r.sym + (r.dedup ? '[DD]' : ''), r.name, '$' + r.price, '24h=' + r.p24.toFixed(2), '7d=' + r.p7.toFixed(2), 'mc=$' + (r.mcap/1e6).toFixed(0) + 'M', 'vmc=' + r.vmc.toFixed(3)].join('\t'));
}

File: .outputs/_score.py
import json, sys

DEDUP = {
    "KAIA","TRAC","BSB","INJ","NEAR","HYPE","CHZ","LIT","VVV","DASH",
    "ZEC","PENGU","GRASS","ONDO","WLD","BEAT","GENIUS","SKYAI","TAG",
    "EIGEN","RAIL"
}

m = json.load(open('.outputs/_markets.json'))
t = json.load(open('.outputs/_trending.json'))
d = json.load(open('.outputs/_dex.json'))

btc = next(c for c in m if c['symbol']=='btc')
eth = next(c for c in m if c['symbol']=='eth')
btc_7d = btc.get('price_change_percentage_7d_in_currency') or 0
eth_7d = eth.get('price_change_percentage_7d_in_currency') or 0
btc_24h = btc.get('price_change_percentage_24h_in_currency') or 0
eth_24h = eth.get('price_change_percentage_24h_in_currency') or 0

print(f"BENCHMARK: BTC ${btc['current_price']} 24h={btc_24h:+.2f}% 7d={btc_7d:+.2f}%")
print(f"BENCHMARK: ETH ${eth['current_price']} 24h={eth_24h:+.2f}% 7d={eth_7d:+.2f}%")
print()

trending_syms = set()
trending_list = []
for c in t.get('coins', []):
    item = c.get('item', {})
    sym = (item.get('symbol') or '').upper()
    trending_syms.add(sym)
    trending_list.append((sym, item.get('name'), item.get('market_cap_rank')))

print("TRENDING (CG):")
for s,n,r in trending_list[:15]:
    print(f"  {s} - {n} (mcap_rank {r})")
print()

# DexScreener trending symbols
dex_syms = set()
for p in d.get('pairs', []):
    bt = p.get('baseToken', {})
    s = (bt.get('symbol') or '').upper()
    if s: dex_syms.add(s)
print(f"DEX trending symbols count: {len(dex_syms)}")
print()

scored = []
for c in m:
    sym = (c.get('symbol') or '').upper()
    if sym in DEDUP: continue
    mcap = c.get('market_cap') or 0
    if mcap < 20_000_000: continue
    vol = c.get('total_volume') or 0
    price = c.get('current_price') or 0
    p24 = c.get('price_change_percentage_24h_in_currency') or 0
    p7 = c.get('price_change_percentage_7d_in_currency') or 0
    vmc = vol / mcap if mcap else 0

    score = 0
    if p24 > 0: score += 1
    if p7 > 0: score += 1
    if p24 > 5 and p7 > 5: score += 2
    if sym in trending_syms: score += 2
    if vmc >= 0.20: score += 3
    elif vmc >= 0.10: score += 2
    if p7 > btc_7d and p7 > eth_7d: score += 2
    if sym in dex_syms: score += 1

    scored.append((score, sym, c.get('name'), price, p24, p7, mcap, vol, vmc,
                   sym in trending_syms, sym in dex_syms))

scored.sort(reverse=True)
print("TOP 20 CANDIDATES:")
print(f"{'sym':<10} {'score':<5} {'24h%':<8} {'7d%':<8} {'vmc':<6} {'mcap':<14} {'trend':<5} {'dex':<5} name")
for s,sym,name,price,p24,p7,mcap,vol,vmc,tr,dx in scored[:20]:
    print(f"{sym:<10} {s:<5} {p24:+7.2f} {p7:+7.2f} {vmc:5.2f} ${mcap/1e9:>9.2f}b   {'Y' if tr else '-':<5} {'Y' if dx else '-':<5} {name}")
